util/src/builder/command.rs

```diff
@@ -0,0 +1,759 @@
+    pub const fn new(name: String, description: String) -> Self {
+        Self(OptionsCommandOptionData {
+            description,
+            name,
+            options: Vec::new(),
+            required: false,
+        })
+    }
+
+    /// Consume the builder, returning the built command option.
+    #[allow(clippy::missing_const_for_fn)]
+    #[must_use = "should be used in a command builder"]
+    pub fn build(self) -> CommandOption {
+        CommandOption::SubCommandGroup(self.0)
+    }
+
+    /// Add a sub command option to the sub command group.
+    ///
+    /// Defaults to an empty list.
+    pub fn option(mut self, option: SubCommandBuilder) -> Self {
+        self.0.options.push(option.into());
+
+        self
+    }
+}
+
+impl From<SubCommandGroupBuilder> for CommandOption {
+    fn from(builder: SubCommandGroupBuilder) -> CommandOption {
+        builder.build()
+    }
+}
+
+/// Create a user option with a builder.
+#[derive(Clone, Debug)]
+#[must_use = "should be used in a command builder"]
+pub struct UserBuilder(BaseCommandOptionData);
+
+impl UserBuilder {
+    /// Create a new default [`UserBuilder`].
+    #[must_use = "builders have no effect if unused"]
+    pub const fn new(name: String, description: String) -> Self {
+        Self(BaseCommandOptionData {
+            description,
+            name,
+            required: false,
+        })
+    }
+
+    /// Consume the builder, returning the built command option.
+    #[allow(clippy::missing_const_for_fn)]
+    #[must_use = "should be used in a command builder"]
+    pub fn build(self) -> CommandOption {
+        CommandOption::User(self.0)
+    }
+
+    /// Set whether this option is required.
+    ///
+    /// Defaults to false.
+    pub const fn required(mut self, required: bool) -> Self {
+        self.0.required = required;
+
+        self
+    }
+}
+
+impl From<UserBuilder> for CommandOption {
+    fn from(builder: UserBuilder) -> CommandOption {
+        builder.build()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use static_assertions::assert_impl_all;
+    use std::fmt::Debug;
+
+    assert_impl_all!(CommandBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(BooleanBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(ChannelBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(IntegerBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(MentionableBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(RoleBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(StringBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(SubCommandBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(SubCommandGroupBuilder: Clone, Debug, Send, Sync);
+    assert_impl_all!(UserBuilder: Clone, Debug, Send, Sync);
+
+    #[test]
+    #[allow(clippy::too_many_lines)]
+    fn construct_command_with_builder() {
+        // from <https://discord.com/developers/docs/interactions/application-commands#example-walkthrough>
```

Comment:
This comment should either be removed (since the test is no longer like the link).
We could create a more comprehensive test for all `Builder` types